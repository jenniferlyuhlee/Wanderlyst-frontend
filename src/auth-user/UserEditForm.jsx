import { useState, useContext } from "react";
import { NavLink } from "react-router-dom";
import WanderlystApi from "../utils/api";
import cleanData from "../helpers/cleanData";
import UserContext from "./UserContext";
import Alert from "../shared/Alert";

/** User Edit Form component
 * Allow users to edit basic information or delete account
 */

function UserEditForm(){
    const { currUser, setCurrUser, logout} = useContext(UserContext);
    
    // states: formData, formErrors, saveConfirmed
    const [formData, setFormData] = useState({
        username: currUser.username,
        password: "",
        firstName: currUser.firstName,
        lastName: currUser.lastName,
        location: currUser.location || "",
        bio: currUser.bio || "",
        profilePic: currUser.profilePic || ""
    })
    const [formErrors, setFormErrors] = useState(null);
    const [saveConfirmed, setSaveConfirmed] = useState(false);
    const [settingErrors, setSettingErrors] = useState(null);

    // general callback to update targeted form data
    function handleChange (e) {
        const { name, value } = e.target;
        // default formData to empty string if value is undefined from currUser
        setFormData(data => ({...data, [name]: value || ""}));
    }

    // form submission handler
    async function handleSubmit(e){
        e.preventDefault();
        let updatedUser;
        try{
            // removes duplicate confirmation password from data
            delete formData.confirmPassword  
            const cleanedFormData = cleanData(formData);
            updatedUser = await WanderlystApi.updateUser(currUser.username, cleanedFormData);

        }
        catch(err){
            setFormErrors(err)
            return;
        }
        // resets profile form to display updated info
        setFormErrors(null);
        setSaveConfirmed(true);
        setCurrUser({
            username: updatedUser.username, 
            firstName: updatedUser.firstName,
            lastName: updatedUser.lastName,
            location: updatedUser.location,
            bio: updatedUser.bio,
            profilePic: updatedUser.profilePic,
            isAdmin: updatedUser.isAdmin, 
            likes: currUser.likes});
    }
    

       // delete user handler
    async function handleDelete(){
        try{
            alert(`Deleted account!`)
            const deleted = await WanderlystApi.deleteUser(currUser.username)
            if (deleted){
                logout();
            }
        }
        catch(err){
            setSettingErrors(err);
        }
    }

    return (
        <div className = "container">
            <NavLink to ={`/users/profile/${currUser.username}`}
            className = "btn rounded-pill btn-secondary mt-4 mb-3">
                Return
            </NavLink>
        <div className = "card px-5 mb-5">
            <h2>Edit User Info</h2>
            <form onSubmit = {handleSubmit} className = "row m-1">
                <div className = "form-floating col-6 mb-2 p-0">
                    <input id = "firstName"
                        className = "form-control"
                        name = "firstName"
                        placeholder = "firstName"
                        value = {formData.firstName}
                        onChange = {handleChange}
                    />
                    <label htmlFor = "firstName">First Name</label>
                </div>
                <div className = "form-floating col-6 mb-2 p-0">
                    <input id = "lastName"
                        className = "form-control"
                        name = "lastName"
                        placeholder = "lastName"
                        value = {formData.lastName}
                        onChange = {handleChange}
                    />
                    <label htmlFor = "lastName">Last Name</label>
                </div>
                <div className = "form-floating col-12 mb-2 p-0">
                    <input id = "location"
                        className = "form-control"
                        name = "location"
                        placeholder = "location"
                        value = {formData.location}
                        onChange = {handleChange}
                    />
                    <label htmlFor = "location">Location</label>
                </div>
                <div className = "form-floating col-12 mb-2 p-0">
                    <input id = "bio"
                        className = "form-control"
                        name = "bio"
                        placeholder = "bio"
                        value = {formData.bio}
                        onChange = {handleChange}
                    />
                    <label htmlFor = "bio">Bio</label>
                </div>
                <div className = "form-floating col-12 mb-2 p-0">
                    <input id = "profilePic"
                        className = "form-control"
                        name = "profilePic"
                        placeholder = "profilePic"
                        value = {formData.profilePic}
                        onChange = {handleChange}
                    />
                    <label htmlFor = "profilePic">Profile Pic URL</label>
                </div>
                <div className = "form-floating col-6 mb-2 p-0">
                    <input id = "username"
                        className = "form-control"
                        name = "username"
                        placeholder = "username"
                        value = {formData.username}
                        onChange = {handleChange}
                        disabled
                    />
                    <label htmlFor = "username">Username</label>
                </div>
                <div className = "form-floating col-6 mb-2 p-0">
                    <input id = "password"
                        className = "form-control"
                        name = "password"
                        type = "password"
                        placeholder = "password"
                        value = {formData.password}
                        onChange = {handleChange}
                    />
                    <label htmlFor = "password">New Password</label>
                </div>
                {formErrors?
                    <Alert type = "danger" messages = {formErrors}/>
                    :
                    null
                }
                <button className = "btn rounded-pill btn-lg btn-primary">
                    Save
                </button>
                {saveConfirmed?
                    <Alert type="success" messages={["Updated successfully."]} />
                    :
                    null
                }
            </form>
            <hr  className = "my-3"/>
            <div>
                <h3>Settings</h3>
                <button onClick={handleDelete} className="btn rounded-pill btn-lg btn-danger">
                    Delete Account
                </button>
                {settingErrors? 
                    <Alert type="warning" messages={["Unable to proceed with your request. Try again later."]} />
                    :
                    null
                }
            </div>
        

        </div>
        </div>
    )
}
export default UserEditForm;