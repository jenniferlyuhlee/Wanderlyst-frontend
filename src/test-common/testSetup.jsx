import React, { useState } from "react";
import UserContext from "../auth-user/UserContext";
import WanderlystApi from "../utils/api";
import '@testing-library/jest-dom';

const demoUser = {
  username: "testadmin", 
  firstName: "testfirst",
  lastName: "testlast",
  location: "testloc",
  bio: "testbio",
  profilePic: null,
  isAdmin: true,
};

const demoTags = [
  {id: 1, name: 'Foodie', description: 'Savor the world one bite at a time. Dive into a journey packed with delicious local cuisines, hidden culinary gems, and must-try dishes that will leave your taste buds tingling.'},
  {id: 2, name: 'Explorer', description: 'Step off the beaten path and uncover the unknown. From ancient ruins to vibrant cityscapes, let your curiosity guide you through unforgettable experiences.'}
]

const UserProvider = ({ children, 
                        currUser = demoUser, 
                        setCurrUser, 
                        tags = demoTags,
                        logout,
                        hasLikedItin,
                        toggleLike }) => {
  const [user, setUser] = useState(currUser);

  return (
    <UserContext.Provider value={{ currUser: user, 
                                  setCurrUser: setUser, 
                                  tags, logout, 
                                  hasLikedItin, toggleLike}}>
      {children}
    </UserContext.Provider>
  );
};

export { UserProvider };