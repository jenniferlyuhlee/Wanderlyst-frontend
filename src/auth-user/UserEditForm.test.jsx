import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { UserProvider } from "../test-common/testSetup";
import { MemoryRouter } from "react-router-dom";
import WanderlystApi from "../utils/api";
import UserEditForm from "./UserEditForm";
import { beforeEach, afterEach, expect, describe, test , vi} from "vitest";

describe("UserEditForm Component", () => {
    // Mock functions passed from context
    const mockSetCurrUser = vi.fn();
    
    // Mock the API methods before running tests
    beforeEach(() => {
        vi.spyOn(WanderlystApi, "updateUser").mockResolvedValue({
            username: "testadmin",
            firstName: "UpdatedFirst",
            lastName: "UpdatedLast",
            location: "UpdatedLocation",
            bio: "UpdatedBio",
            profilePic: "UpdatedPic",
            isAdmin: false
        });

        vi.spyOn(WanderlystApi, "deleteUser").mockResolvedValue(true);
    });

    // Restore mocks after tests
    afterEach(() => {
        vi.restoreAllMocks();
    });
 
    test("renders without crashing", () => {
        render(
        <MemoryRouter>
            <UserProvider setCurrUser = {mockSetCurrUser}>
                <UserEditForm />
            </UserProvider>
        </MemoryRouter>)
    });
            
    
    test("matches snapshot", () => {
        const {asFragment} = render(
                            <MemoryRouter>
                                <UserProvider setCurrUser = {mockSetCurrUser}>
                                    <UserEditForm />
                                </UserProvider>
                            </MemoryRouter>)
        expect(asFragment()).toMatchSnapshot();
    });

    test("calls updateUser on form submission", async () => {
        const {getByLabelText, getByText} = render(
            <MemoryRouter>
                <UserProvider setCurrUser = {mockSetCurrUser}>
                    <UserEditForm />
                </UserProvider>
            </MemoryRouter>);

        // capture fields and login button
        const firstName = getByLabelText("First Name");
        const lastName = getByLabelText("Last Name");
        const location = getByLabelText("Location");
        const bio = getByLabelText("Bio");
        const profilePic = getByLabelText("Profile Pic URL");
        const btn = getByText("Save");

        // fill in fields and submit
        fireEvent.change(firstName, {target: {value: 'updatedfirst'}});
        fireEvent.change(lastName, {target: {value: 'updatedlast'}});
        fireEvent.change(location, {target: {value: 'updatedlocation'}});
        fireEvent.change(bio, {target: {value: 'updatedbio'}});
        fireEvent.change(profilePic, {target: {value: 'https://updatedPic.jpg'}});
        fireEvent.click(btn);

        // waits for mock func to be called
        await waitFor(() => {
            expect(WanderlystApi.updateUser).toHaveBeenCalledWith("testadmin",
                {
                        username: "testadmin",
                        firstName: "updatedfirst",
                        lastName: "updatedlast",
                        location: "updatedlocation",
                        bio: "updatedbio",
                        profilePic: "https://updatedPic.jpg",
                    }
            );
        });
        // Check if success alert appears
        await waitFor(() => {
            expect(getByText("Updated successfully.")).toBeInTheDocument();
        })
    });

    test("calls deleteUser when delete button clicked", async () => {
        const {getByText} = render(
            <MemoryRouter>
                <UserProvider>
                    <UserEditForm />
                </UserProvider>
            </MemoryRouter>);

        // Click delete button
        fireEvent.click(getByText("Delete Account"));

        // wait for mock func to be called
        await waitFor(() => {
            expect(WanderlystApi.deleteUser).toHaveBeenCalledWith("testadmin");
        });
    })
})