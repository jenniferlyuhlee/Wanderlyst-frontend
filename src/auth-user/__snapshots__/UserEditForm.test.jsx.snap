// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserEditForm Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <a
      class="btn rounded-pill btn-secondary mt-4 mb-3"
      href="/users/profile/testadmin"
    >
      Return
    </a>
    <div
      class="card px-5 mb-5"
    >
      <h2>
        Edit User Info
      </h2>
      <form
        class="row m-1"
      >
        <div
          class="form-floating col-6 mb-2 p-0"
        >
          <input
            class="form-control"
            id="firstName"
            name="firstName"
            placeholder="firstName"
            value="testfirst"
          />
          <label
            for="firstName"
          >
            First Name
          </label>
        </div>
        <div
          class="form-floating col-6 mb-2 p-0"
        >
          <input
            class="form-control"
            id="lastName"
            name="lastName"
            placeholder="lastName"
            value="testlast"
          />
          <label
            for="lastName"
          >
            Last Name
          </label>
        </div>
        <div
          class="form-floating col-12 mb-2 p-0"
        >
          <input
            class="form-control"
            id="location"
            name="location"
            placeholder="location"
            value="testloc"
          />
          <label
            for="location"
          >
            Location
          </label>
        </div>
        <div
          class="form-floating col-12 mb-2 p-0"
        >
          <input
            class="form-control"
            id="bio"
            name="bio"
            placeholder="bio"
            value="testbio"
          />
          <label
            for="bio"
          >
            Bio
          </label>
        </div>
        <div
          class="form-floating col-12 mb-2 p-0"
        >
          <input
            class="form-control"
            id="profilePic"
            name="profilePic"
            placeholder="profilePic"
            value=""
          />
          <label
            for="profilePic"
          >
            Profile Pic URL
          </label>
        </div>
        <div
          class="form-floating col-6 mb-2 p-0"
        >
          <input
            class="form-control"
            disabled=""
            id="username"
            name="username"
            placeholder="username"
            value="testadmin"
          />
          <label
            for="username"
          >
            Username
          </label>
        </div>
        <div
          class="form-floating col-6 mb-2 p-0"
        >
          <input
            class="form-control"
            id="password"
            name="password"
            placeholder="password"
            type="password"
            value=""
          />
          <label
            for="password"
          >
            New Password
          </label>
        </div>
        <button
          class="btn rounded-pill btn-lg btn-primary"
        >
          Save
        </button>
      </form>
      <hr
        class="my-3"
      />
      <div>
        <h3>
          Settings
        </h3>
        <button
          class="btn rounded-pill btn-lg btn-danger"
        >
          Delete Account
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
